src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,422 @@
+`slash_invariant_forms_class` by requiring that the functions be holomorphic and zero
+at infinity. -/
+class cusp_form_class extends slash_invariant_form_class F Γ k :=
+(hol : ∀ f : F, mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (f : ℍ → ℂ))
+(zero_at_infty : ∀ (f : F) (A : SL(2, ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+@[priority 100]
+instance modular_form_class.modular_form : modular_form_class (modular_form Γ k) Γ k :=
+{ coe := modular_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := modular_form.slash_action_eq',
+  hol := modular_form.hol',
+  bdd_at_infty := modular_form.bdd_at_infty' }
+
+@[priority 100]
+instance cusp_form_class.cusp_form : cusp_form_class (cusp_form Γ k) Γ k :=
+{ coe := cusp_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := cusp_form.slash_action_eq',
+  hol := cusp_form.hol',
+  zero_at_infty := cusp_form.zero_at_infty' }
+
+variables {F Γ k}
+
+instance : has_coe_to_fun (modular_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+instance : has_coe_to_fun (cusp_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma modular_form_to_fun_eq_coe {f : modular_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+@[simp] lemma cusp_form_to_fun_eq_coe {f : cusp_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem modular_form_ext {f g : modular_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+@[ext] theorem cusp_form_ext {f g : cusp_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `modular_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def modular_form_copy (f : modular_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  modular_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  bdd_at_infty' := λ A, h.symm ▸ f.bdd_at_infty' A }
+
+/-- Copy of a `cusp_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def cusp_form_copy (f : cusp_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  cusp_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_forms
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { to_fun := f + g,
+    slash_action_eq' := (f.to_slash_invariant_form + g.to_slash_invariant_form).2,
+    hol' := f.hol'.add _ g.hol',
+    bdd_at_infty' := λ A, begin
+        rw slash_action.add_action,
+        exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (f.bdd_at_infty' A) (g.bdd_at_infty' A),
+        end}}
+
+@[simp] lemma coe_add {f g : (modular_form Γ k)} : ⇑(f + g) = f + g := rfl
+@[simp] lemma add_apply {f g : (modular_form Γ k)} {z : ℍ} : (f + g) z = f z + g z := rfl
```

Comment:
I don't think you need the `apply` version; the `coe` version should cover it.  Same comment on other `apply`/`coe` pairs throughout.

Reply:
It's usually convenient to have both, and we do so elsewhere; `rw [foo.coe_add, pi.add_apply]` is annoying. So I'd argue for keeping both.

Reply:
But the arguments should be explicit (throughout):
```suggestion
@[simp] lemma add_apply (f g : (modular_form Γ k)) (z : ℍ) : (f + g) z = f z + g z := rfl
```

Reply:
I've made them explicit. I happy to keep/remove it so whatever the majority want we can go with.